client.go

```diff
@@ -1355,7 +1367,47 @@ func (c *HostClient) acquireConn() (*clientConn, error) {
 		return cc, nil
 	}
 	if !createConn {
-		return nil, ErrNoFreeConns
+		if c.MaxConnWaitTimeout <= 0 {
+			return nil, ErrNoFreeConns
+		}
+
+		// reqTimeout    c.MaxConnWaitTimeout   wait duration
+		//     d1                 d2            min(d1, d2)
+		//  0(not set)            d2            d2
+		//     d1            0(don't wait)      0(don't wait)
+		//  0(not set)            d2            d2
+		timeout := c.MaxConnWaitTimeout
+		timeoutOverridden := false
+		// reqTimeout == 0 means not set
+		if reqTimeout > 0 && reqTimeout < timeout {
+				timeout = reqTimeout
+				timeoutOverridden = true
+		}
+
+		// wait for a free connection
+		tc := AcquireTimer(timeout)
+		defer ReleaseTimer(tc)
+
+		w := &wantConn{
+			ready: make(chan struct{}, 1),
+		}
```

Comment:
Would it be possible to reuse `wantConn` or the channel? Or is that going to make the code so complicated that its not worth it?

Reply:
For the first question, if we want to reuse the `wantConn`, then we must reuse the channel, but AFAIK there's no way to reopen a closed channel (maybe).
In terms of the heap allocation, yes, this will cause heap allocation and slow things down, but if we want to waiting for a free connection, the waiting dominates the slow, not gc.

For the second question, many http clients provide connection pooling and waiting mechanism (by blocking). This waiting mechanism is very useful, it can simplify the code on the caller side, otherwise the caller have two options:

1. busy waiting, try again immediately at some max try count, which consumes cpu time.
2. sleep waiting, sleep for a while and try again, which can't get connection in time.

Both are not perfect. So I think the waiting mechanism is worth it.

For the complication of this PR, em ..., actually this PR is based on [net/http/transport.go][1], maybe there is a more simple way, we can have a discussion about that :)

[1]: https://github.com/golang/go/blob/master/src/net/http/transport.go#L95

Reply:
I have thought about it and I agree. Let's merge this!